crypto/sm2/sm2_pmeth.c

```diff
@@ -155,29 +171,51 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
             SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_INVALID_CURVE);
             return 0;
         }
-        EC_GROUP_free(dctx->gen_group);
-        dctx->gen_group = group;
+        EC_GROUP_free(smctx->gen_group);
+        smctx->gen_group = group;
         return 1;
 
     case EVP_PKEY_CTRL_EC_PARAM_ENC:
-        if (dctx->gen_group == NULL) {
+        if (smctx->gen_group == NULL) {
             SM2err(SM2_F_PKEY_SM2_CTRL, SM2_R_NO_PARAMETERS_SET);
             return 0;
         }
-        EC_GROUP_set_asn1_flag(dctx->gen_group, p1);
+        EC_GROUP_set_asn1_flag(smctx->gen_group, p1);
         return 1;
 
     case EVP_PKEY_CTRL_MD:
-        dctx->md = p2;
+        smctx->md = p2;
         return 1;
 
     case EVP_PKEY_CTRL_GET_MD:
-        *(const EVP_MD **)p2 = dctx->md;
+        *(const EVP_MD **)p2 = smctx->md;
+        return 1;
+
+    case EVP_PKEY_CTRL_SM2_SET_ID:
```

Comment:
Probably this should be `...SET1_ID`?

Reply:
Yes. The number is supposed to give you a hint about the memory management:

get0 == returns a pointer to the internal data. The caller should therefore not free it
get1 == creates a copy of the data or ups a ref count. The caller should free.
set0 == the caller passes ownership of the pointer to the library, so the caller should not subsequently free it
set1 == the caller does not pass ownership of the pointer to the library. The library will take a copy or up a ref count. The caller should free the original pointer.

Reply:
Understood, thanks for the clear explanation.